---
title: "Computational Biology Course Summary SoSe21"
author: Finn Lo
output:
  html_document:
    toc: yes
    number_section: no
    fig_caption: true
  pdf_document:
    toc: yes
  always_allow_html: true
date: "`r format(Sys.time(), '%d %B %Y')`"
---

# CHAPTER 4 - RIEMER ET AL. 2017 - RNA-SEQ

## 4.1 Loading libraries and data (Riemer)

``` {r}
library(matrixStats)
library(GO.db)
library(pheatmap)
library(DESeq2)
library(goseq)
library(tidyverse)
library(ggplot2)
library(tidyr)
library(stringr)

Expression_Data <- read_delim("https://raw.githubusercontent.com/tlobnow/Resources/main/R_Introduction/CompBio_Data/expression_matrix.tab", col_names = F, delim = "\t")
Exp_Design      <- read_delim("https://raw.githubusercontent.com/tlobnow/Resources/main/R_Introduction/CompBio_Data/samples.tsv", delim = "\t", col_names = T)
Annotations     <- read_delim("https://raw.githubusercontent.com/tlobnow/Resources/main/R_Introduction/CompBio_Data/annotations.tsv", delim = " ")
```

## 4.2 Data Exploration
``` {r}
#### How is it organized?
glimpse(Expression_Data)
#dim(Expression_Data)
#View(Expression_Data)
arrange(Expression_Data)

glimpse(Exp_Design)
#dim(Exp_Design)
#View(Exp_Design)
arrange(Exp_Design)

glimpse(Annotations)
#dim(Annotations)
#View(Annotations)
arrange(Annotations)

#### For how many genes do we have expression data?
# 47643 genes, 64884
#### How many samples are there?
# 12 columns w/ samples? 6 samples
#### How many replicates?
# 2 replicates

#### Reorganise data and provide sensible names for the columns:
colnames(Expression_Data) <- make.unique(c("ensg", Exp_Design$oncogene))

Expression_Data %>%
  dplyr::select("KRAS", "KRAS.1", 
                "PIK3CA+CTNNB1", "PIK3CA+CTNNB1.1", 
                "PIK3CA", "PIK3CA.1", 
                "CTNNB1", "CTNNB1.1", 
                "BRAF", "BRAF.1", 
                "cntrl", "cntrl.1")

#### Annotate the columns of expression matrix using the 
#### gene names from the annotation file --> simply add a new column!

data <- Annotations %>%
  dplyr::select(-transcript_type) %>%
  distinct() %>%
  right_join(Expression_Data)


#data <- Expression_Data %>%
#  left_join(Annotations)

#### the data == Ribo-minus ... total RNA w/o ribosomal RNA (should be depleted)
#### how successful was that?
#### what is annotated?

geneIDsRRNA <- Annotations %>%
  filter(str_detect(transcript_type, "rRNA")) %>%
  pull(ensg) %>%
  unique()


data_to_plot <- data %>%
  filter(ensg %in% geneIDsRRNA) %>%
  dplyr::select(-ensg, -gene_name)

pheatmap(data_to_plot, show_rownames = F)
```

``` {r}
f_data_to_plot <- data_to_plot %>% filter(rowSums(.)>0)

pheatmap(f_data_to_plot, show_rownames = F)
```

``` {r}
#### there are only 3 abundant rRNA genes, big diff. between samples
#### --> remove all rRNAs from the data

data <- data %>%
  filter(!ensg %in% geneIDsRRNA)

#### use ggplot boxplots/histograms
####    histograms:  good for more detailed understanding of single-sample  distributions
####    boxplots:    good for comparing samples

#### Data pivoted
data_piv <- data %>%
  pivot_longer(cols = c("KRAS",  
                        "PIK3CA+CTNNB1", 
                        "PIK3CA+CTNNB1.1", 
                        "PIK3CA", 
                        "CTNNB1", 
                        "BRAF", 
                        "PIK3CA.1", 
                        "KRAS.1",
                        "CTNNB1.1", 
                        "cntrl", 
                        "cntrl.1",
                        "BRAF.1"
                        ),
               values_to = "value",
               names_to = "Sample"
               )

#### all Samples as histograms, wrapped
data_piv %>%
  ggplot(aes(value + 1)) +
  geom_histogram() +
  facet_wrap(~Sample)
```

``` {r}
data_piv %>%
  ggplot(aes(log2(value + 1), fill = Sample)) +
  geom_histogram() +
  facet_wrap(~Sample)
```

``` {r}
#### all Samples as Boxplots
data_piv %>%
  ggplot(aes(log2(value + 1), Sample, col = Sample)) +
  geom_boxplot()
```

``` {r}
#### geom_density
data_piv %>%
  ggplot(aes(log2(value + 1), col = Sample)) +
  geom_density() +
  facet_wrap(~Sample)
```

``` {r}
#### geom_violin
data_piv %>%
  ggplot(aes(log2(value + 1), Sample, col = Sample)) +
  geom_violin()

```

``` {r}
librarySize <- data %>%
  summarise(across(where(is.numeric),sum)) %>%
  pivot_longer(everything(), names_to = "Sample",
               values_to = "size")
```
## 4.3. Normalization of the RNAseq data

``` {r}
#### use RPM (reads per million)
rpm <- data %>%
  mutate_if(is.numeric, ~. / (sum(.) / 10^6))

#### test whether the sum of each column has the same value
rpm$KRAS %>% sum()
rpm$BRAF %>% sum()

library(DESeq2)

#### Estimating Library Size --> DESeq2 (needs a specific format)
deseq_Data <- DESeqDataSetFromMatrix(countData = data[ ,3:14],
                                     colData = Exp_Design,
                                     design = ~oncogene)
feature_Data <- data.frame(gene = data$ensg)                                     
mcols(deseq_Data) <- DataFrame(mcols(deseq_Data), feature_Data)

#### get rid of all genes with no reads:
deseq_Data <- deseq_Data[ rowSums(counts(deseq_Data)) > 1,]
deseq_Data <- estimateSizeFactors(deseq_Data)

sizeFactors(deseq_Data)

libsize_Estimates <- tibble(read_count = librarySize$size,
                            deseq_Lib_Size = sizeFactors(deseq_Data))
#### library size estimates based on reads per miilion
#### correlates with deseq2, but they are NOT IDENTICAL
libsize_Estimates %>%
  ggplot(aes(read_count, deseq_Lib_Size)) +
  geom_point()
```

``` {r}
#### look at deseq2-normalized data:
norm_Counts <- deseq_Data %>%
  counts(normalized = T) %>%
  as_tibble()

rownames(norm_Counts) <- deseq_Data %>%
  mcols %>%
  .$gene

reform_data <- norm_Counts %>%
                  pivot_longer(everything(),
                               names_to = "Sample",
                               values_to = "value")

reform_data %>%
  ggplot(aes(x = value, fill = Sample)) +
  geom_density(alpha = 0.2) +
  scale_x_log10() +
  geom_vline(aes(xintercept = 75))
```

``` {r}
reform_data %>%
  ggplot(aes(log2(value), fill = Sample)) +
  geom_histogram() +
  facet_wrap(~Sample)
#### now the histograms look much more similar
```

``` {r}
#### we want to make a subset of data with EXPRESSED genes only
#### make cut-off at 75 counts (== "dip" in the histogram)
#### we only look at samples expressed over 75 in at least 2 samples
#### == selected normalized count
sel_norm_count <- norm_Counts %>%
  filter(rowSums(.>75)>1)

#### check
sel_norm_count %>% dim()
```

## 4.4 Analyse
``` {r}
pca <- sel_norm_count %>%
  t() %>%
  prcomp()

PCA_sdev <- tibble(sdev = pca$sdev,
                   PC = pca$x %>% colnames) %>%
  mutate(var_exp = sdev^2 / (sum(sdev^2))) %>% # var_exp == Variance explained
  mutate(PC = factor(PC, levels = PC))

PCA_sdev %>%
  ggplot(aes(PC, var_exp)) +
  geom_col()
#### shows the variances (y axis) associated w/ PCs (x axis)
#### useful for deciding how many PCs to retain for further analysis
#### here the first two explain most variability
#### now we can plot them
```

## 4.4 Analyse
``` {r}
pca_X <- pca$x %>%
  as_tibble() %>%
  dplyr::select(PC1, PC2) %>%
  mutate(Sample = rownames(pca$x))

pca_X %>%
  ggplot(aes(PC1, PC2, label = Sample)) +
  geom_label()
#### in RNA-Seq the var grows with the mean
#### e.g. f one performs PCA directly on a matrix of normalized read counts, 
#### the result typically depends only on the few most strongly expressed genes 
#### because they show the largest absolute differences between samples. 
#### As a solution, DESeq2 offers the regularized–logarithm transformation, 
#### or RLOG for short
```

``` {r}
#### redo the PCA on rlog transformed data, by using DESeq2 rlogTransformation()
#### changes?
#### use plotPCA() instead of ggplot() 
#### --> specifically designed to create / plot data from deseq2 dataset

#?rlogTransform

r_deseq_Data <- rlogTransformation(deseq_Data)

plotPCA(r_deseq_Data, intgroup = "oncogene") # how to specify colors

#### In both cases, most samples cluster on the left side, 
#### and the two BRAF samples are in a different side, and are different 
#### mainly due to PC1. This reflects the biology of the system, 
#### as BRAF is a very strong oncogene and eliminates basically all stem cells 
#### in the culture. In the second PCA the replicates always cluster together, 
#### except for one KRAS sample. Note that the PC1 already covers 84% of the variance. 
#### To investigate this further, let’s have a look at the 1000 most varying genes 
#### in the data set:
```

``` {r}
most_var_genes <- sel_norm_count %>%
  mutate(sd = as.matrix(.) %>%
           rowSds()) %>%
  arrange(-sd) %>%
  dplyr::slice(1:1000) %>% ### index rows by their integer locations, can also be slice_head
  dplyr::select(-sd) %>%
  as.matrix()

pheatmap(most_var_genes, show_rownames = F)  
#### we don't see much.. a few genes have a much higher expression governing the color scale
#### what happens, when the rows are scaled:
```

``` {r}
pheatmap(most_var_genes, show_rownames = F, scale = "row")
#### Much better!
```

``` {r}
#### 1000 most highly expressed genes?
most_high_genes <- sel_norm_count %>%
  mutate(mean = as.matrix(.) %>%
           rowMeans2()) %>%
  arrange(desc(mean)) %>%
  dplyr::slice(1:1000) %>% ### index rows by their integer locations, can also be slice_head
  dplyr::select(-mean) %>%
  as.matrix()

#### in reality one would filter for genes with median in top50% and then
#### at the 1000 highest varying genes
#### any changes in clustering?

ggplot(log2(sel_norm_count+1), aes(cntrl, cntrl.1)) +
  geom_point(size = 0.5)
```

``` {r} 
ggplot(log2(sel_norm_count+1), aes(cntrl, BRAF)) +
  geom_point(size = 0.5)
```

``` {r}
ggplot(log2(sel_norm_count+1), aes(BRAF, BRAF.1)) +
  geom_point(size = 0.5)  
```

``` {r}
ggplot(log2(sel_norm_count+1), aes(BRAF, KRAS)) +
  geom_point(size = 0.5)  

```

``` {r}
#### look at the correlation b/w diff. samples

pheatmap(cor(log2(sel_norm_count+1)))
#### Here, you can note that the BRAF samples are very dissimilar from the rest,
#### and that one of the KRAS samples seems to be an outlier which corresponds 
#### to the results of the PCA (with rlog).
#### To test differential expression, it is important to know how “noisy” 
#### the expression of a gene is. To estimate this, DESeq2 fits a 
#### model of dispersion (variance divided by mean) to the replicates, which we can then plot.
```

``` {r}
deseq_Data <- estimateDispersions(deseq_Data)
plotDispEsts(deseq_Data)
#### First, DESeq2 calculates the dispersion for each gene (black spots). 
#### Then it fits a model to it (red line), and then calculates for each gene a 
#### mix of the gene-wise dispersion and the model-derived dispersion (blue dots). 
#### Genes that show unusually high dispersion are identified as outliers (blue circles) 
#### and will be assigned the gene-wise dispersion (Love, Huber, and Anders 2014).
#### The most often applied analysis for gene expression data is the calculation 
#### of differential expression. Let’s derive the genes that behave differently 
#### between oncogenes and the control.
```

``` {r}
output <- DESeq(deseq_Data)
resultsNames(output)
#### DESeq2 has now calculated differential expressions for all listed comparisons
#### How does it compare?
#### 
#### How does it calculate? --> Manual
```

``` {r}
#### now we can look at the Comparisons:

CTNNB1_vs_Cntrl <- results(output, contrast = c("oncogene", "CTNNB1", "cntrl"))
summary(CTNNB1_vs_Cntrl)
sum(CTNNB1_vs_Cntrl$padj < 0.1, na.rm = T)

#### An MA plot visualizes differences between two groups relative to average signal intensity. 
#### This is useful to assess potential biases in the data. Typically, 
#### most peak intensities aren't expected to change between conditions. 
#### This means points in the plot should be grouped around a horizontal line at 0, 
#### corresponding to no change. MA plots are also useful to get a quick visual impression 
#### of how large the differences between groups are.
#### MA == plot of log-intensity ratios (M-values) vs. log-intensity averages (A-values)

plotMA(CTNNB1_vs_Cntrl, main = "MA Plot", ylim = c(-4,4))
```

``` {r}
#### now let's plot individual genes -
#### -> select one gene w/ lowest adjusted p-value compared to b-Catenin and Ctrl:

plotCounts(output, gene = which.min(CTNNB1_vs_Cntrl$padj), intgroup = "oncogene")
```

``` {r}
CTNNB1_genes <- deseq_Data %>%
  mcols() %>%
  as_tibble() %>%
  dplyr::select(gene) %>%
  filter(CTNNB1_vs_Cntrl$padj < 0.05)

CTNNB1_gene <- deseq_Data %>%
  mcols() %>%
  as_tibble() %>%
  dplyr::select(gene) %>%
  filter(CTNNB1_vs_Cntrl$padj < 0.0000000000000000000000000001) %>%
  pull

CTNNB1_vs_Cntrl_tibble <- CTNNB1_vs_Cntrl %>%
  as_tibble() %>%
  mutate(min_padj = min(padj, na.rm = T))

CTNNB1_gene2 <- deseq_Data %>%
  mcols() %>%
  as_tibble() %>%
  dplyr::select(gene) %>%
  filter(CTNNB1_vs_Cntrl_tibble$padj == CTNNB1_vs_Cntrl_tibble$min_padj) %>%
  pull

gene <- rpm %>%
  filter(ensg == "ENSMUSG00000049350")

###### COMPLETE! which gene is that?
###### use ggplot() for similar plot group vs. normalized count
norm_Counts %>%
  rownames_to_column("ensg") %>%
  filter(ensg %in% gene$ensg) %>%
  pivot_longer(names_to = "group",
               values_to = "normalizedcount", -ensg) %>%
  mutate(group = str_replace(group, pattern = "CTNNB1.1", replacement = "CTNNB1")) %>%
  mutate(group = str_replace(group, pattern = "BRAF.1", replacement = "BRAF")) %>%
  mutate(group = str_replace(group, pattern = "PIK3CA.1", replacement = "PIK3CA")) %>%
  mutate(group = str_replace(group, pattern = "PIK3CA+CTNNB1.1", replacement = "PIK3CA+CTNNB1")) %>%
  mutate(group = str_replace(group, pattern = "cntrl.1", replacement = "cntrl")) %>%
  mutate(group = str_replace(group, pattern = "KRAS.1", replacement = "KRAS")) %>%
  mutate(group = factor(group, levels = c("BRAF", 
                                          "CTNNB1",
                                          "KRAS",
                                          "PIK3CA",
                                          "PIK3CA+CTNNB1",
                                          "cntrl"))) %>%
  ggplot(aes(group, normalizedcount)) +
  geom_point(size = 2, alpha = 0.5, shape = 1) +
  scale_y_log10() +
  theme_bw() +
  ggtitle(gene$gene_name)
```

``` {r, message = F, warning = F}
## 4.5 GO-Enrichment

#deseq_Data <- readRDS('deseq_Data.rds')
#### take genes that are differently expressed in BRAF mutant --> try to characterize better
#### calculate GO-Term enrichment for selected GO-terms, library(Bioconductor) == useful
#### prepare data for the analysis:

#allGenes <- mcols(deseq_Data)$gene
#gene_Vector <- as.integer(allGenes %in% CTNNB1_genes)
#names(gene_Vector) <- allGenes

#### GO Analysis ~~ tests categorical membership vs. hypergeometric null distribution
#### however: null distribution has to be adjusted for RNA-seq data:
#### must be adjusted by the expected length bias --> use probability weighting fct. (PWF)
#### PWF == fct. that gives the probability that a gene will be differentially expressed (based on length only)
#### use nullp() from library(goseq)
#### use annotation package library(geneLenDataBase) to account for length bias of genes!

#library(geneLenDataBase)
#library(goseq)
#supportedGenomes() %>%
#  as_tibble() %>%
#  filter(species == "Mouse")
#### Problem: our genome is mapped to version mm10 


#### --> use Bioconductor package TxDb.Mmusculus.UCSC.mm10.ensGene
#library(BiocManager)
#install("TxDb.Mmusculus.UCSC.mm10.ensGene", force = T)
#library(TxDb.Mmusculus.UCSC.mm10.ensGene)
#pwf <- nullp(gene_Vector, "mm10", "ensGene")
#### Can't find mm10/ensGene length data in genLenDataBase...
#### Found the annotation package, TxDb.Mmusculus.UCSC.mm10.ensGene
#### Trying to get the gene lengths from it.
#### Error in if (matched_frac == 0) { : missing value where TRUE/FALSE needed


#### proceeding with "mm9" for now.
#library(geneLenDataBase)
#library(BiocManager)
#library(tidyverse)
#install.packages(TxDb.Mmusculus.UCSC.mm9.ensGene)
#library(TxDb.Mmusculus.UCSC.mm9.ensGene)

#supportedGenomes() %>%
#  as_tibble() %>%
#  filter(species == "Mouse")

#pwf <- nullp(gene_Vector, "mm10", "ensGene")

#GO_Analysis_BP <- goseq(pwf, "mm10", "ensGene", test.cats = "GO:BP")

#Go_Enriched_BP <- GO_Analysis_BP %>%
                  as_tibble() %>%
                  mutate(p.adjust = p.adjust(over_represented_pvalue, method = "BH")) %>%
                  filter(p.adjust < 0.01)


#### now: significant categories collected..
#### use GO-Terms from GO.db for more detailed descriptions of the GO-Terms
#GOTERM[[ GoEnrichedBP$category[1] ]]

#### Compare GO-Enrichment w/o accounting for length bias --> method = "Hypergeometric"
#GO_Analysis_BP_hyper <- goseq(pwf, "mm10", "ensGene", test.cats = "GO:BP", method = "Hypergeometric")
#Go_Enriched_BP_hyper <- GO_Analysis_BP_hyper %>%
#  as_tibble() %>%
#  mutate(p.adjust = p.adjust(over_represented_pvalue, method = "BH")) %>%
#  filter(p.adjust < 0.01)

#diff <- setdiff(Go_Enriched_BP, Go_Enriched_BP_hyper)

#### repeat analysis for GO-categories: "Cellular Compartment"
####                                    "Molecular Function"
```